File: api/client-samples/php/SdkLike.php
<?php

/**
 * NameDivider API PHP SDK
 * 
 * このクラスは、NameDivider APIのPHP SDKの実装です。
 * 基本的な使用方法から、異なるアルゴリズムモードの使い分けまで対応しています。
 */

/**
 * 分割された名前の結果を表すクラス
 */
class DividedName
{
    public string $family;
    public string $given;
    public string $separator;
    public float $score;
    public string $algorithm;

    public function __construct(array $data)
    {
        $this->family = $data['family'];
        $this->given = $data['given'];
        $this->separator = $data['separator'];
        $this->score = $data['score'];
        $this->algorithm = $data['algorithm'];
    }

    public function __toString(): string
    {
        return sprintf(
            "DividedName(family='%s', given='%s', score=%.4f, algorithm='%s')",
            $this->family,
            $this->given,
            $this->score,
            $this->algorithm
        );
    }
}

/**
 * NameDivider API クライアント
 */
class NameDividerClient
{
    private string $baseUrl;
    private array $curlOptions;

    /**
     * コンストラクタ
     * 
     * @param string $baseUrl APIのベースURL
     */
    public function __construct(string $baseUrl = 'http://localhost:8000')
    {
        $this->baseUrl = $baseUrl;
        $this->curlOptions = [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => ['Content-Type: application/json'],
            CURLOPT_TIMEOUT => 30,
            CURLOPT_CONNECTTIMEOUT => 10,
        ];
    }

    /**
     * 名前を分割する
     * 
     * @param array<string> $names 分割する名前の配列
     * @param string $mode 分割モード ('basic' または 'gbdt')
     * @return array<DividedName> 分割結果の配列
     * @throws Exception APIエラーまたはネットワークエラーの場合
     */
    public function divide(array $names, string $mode = 'basic'): array
    {
        $request = [
            'names' => $names,
            'mode' => $mode
        ];

        $ch = curl_init();
        curl_setopt_array($ch, $this->curlOptions + [
            CURLOPT_URL => $this->baseUrl . '/divide',
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request),
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($response === false) {
            throw new Exception("cURL error: $error");
        }

        if ($httpCode !== 200) {
            throw new Exception("HTTP error: $httpCode");
        }

        $data = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new Exception('Invalid JSON response: ' . json_last_error_msg());
        }

        if (!isset($data['divided_names'])) {
            throw new Exception('Invalid response format');
        }

        return array_map(
            fn($item) => new DividedName($item),
            $data['divided_names']
        );
    }

    /**
     * BasicNameDividerを使用して名前を分割
     * 
     * @param array<string> $names 分割する名前の配列
     * @return array<DividedName> 分割結果の配列
     */
    public function divideBasic(array $names): array
    {
        return $this->divide($names, 'basic');
    }

    /**
     * GBDTNameDividerを使用して名前を分割
     * 
     * @param array<string> $names 分割する名前の配列
     * @return array<DividedName> 分割結果の配列
     */
    public function divideGbdt(array $names): array
    {
        return $this->divide($names, 'gbdt');
    }
}

File: api/client-samples/kotlin/src/main/kotlin/SdkLike.kt
import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * 分割された名前の結果を表すデータクラス
 */
@Serializable
data class DividedName(
    val family: String,
    val given: String,
    val separator: String,
    val score: Double,
    val algorithm: String
) {
    override fun toString(): String {
        return "DividedName(family='$family', given='$given', score=${String.format("%.4f", score)}, algorithm='$algorithm')"
    }
}

/**
 * API リクエストのペイロード
 */
@Serializable
data class DivideRequest(
    val names: List<String>,
    val mode: String = "basic"
)

/**
 * API レスポンスの構造
 */
@Serializable
data class DivideResponse(
    @SerialName("divided_names")
    val dividedNames: List<DividedName>
)

/**
 * NameDivider API Kotlin SDK
 * 
 * このクラスは、NameDivider APIのKotlin SDKの実装です。
 * 基本的な使用方法から、異なるアルゴリズムモードの使い分けまで対応しています。
 * Kotlin Coroutinesを使用した非同期処理をサポートしています。
 */
class NameDividerClient(
    private val baseUrl: String = "http://localhost:8000",
    private val timeout: Duration = Duration.ofSeconds(30)
) {
    private val httpClient = HttpClient.newBuilder()
        .connectTimeout(timeout)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    /**
     * 名前を分割する
     * 
     * @param names 分割する名前のリスト
     * @param mode 分割モード ('basic' または 'gbdt')
     * @return 分割結果のリスト
     * @throws IllegalArgumentException 引数が無効な場合
     * @throws Exception API通信エラーの場合
     */
    suspend fun divide(names: List<String>, mode: String = "basic"): List<DividedName> = withContext(Dispatchers.IO) {
        require(names.isNotEmpty()) { "Names cannot be empty" }
        require(mode in listOf("basic", "gbdt")) { "Mode must be 'basic' or 'gbdt'" }

        val request = DivideRequest(names, mode)
        val requestBody = json.encodeToString(request)

        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseUrl/divide"))
            .header("Content-Type", "application/json")
            .timeout(timeout)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody))
            .build()

        val response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString())

        if (response.statusCode() !in 200..299) {
            throw Exception("HTTP error: ${response.statusCode()}")
        }

        try {
            val divideResponse = json.decodeFromString<DivideResponse>(response.body())
            divideResponse.dividedNames
        } catch (e: SerializationException) {
            throw Exception("Failed to parse response: ${e.message}", e)
        }
    }

    /**
     * BasicNameDividerを使用して名前を分割
     * 
     * @param names 分割する名前のリスト
     * @return 分割結果のリスト
     */
    suspend fun divideBasic(names: List<String>): List<DividedName> = divide(names, "basic")

    /**
     * GBDTNameDividerを使用して名前を分割
     * 
     * @param names 分割する名前のリスト
     * @return 分割結果のリスト
     */
    suspend fun divideGbdt(names: List<String>): List<DividedName> = divide(names, "gbdt")

    /**
     * 複数のモードを並列実行してアルゴリズムを比較
     * 
     * @param names 分割する名前のリスト
     * @return BasicとGBDTの結果をPairで返す
     */
    suspend fun compareAlgorithms(names: List<String>): Pair<List<DividedName>, List<DividedName>> = coroutineScope {
        val basicDeferred = async { divideBasic(names) }
        val gbdtDeferred = async { divideGbdt(names) }
        
        Pair(basicDeferred.await(), gbdtDeferred.await())
    }

    /**
     * バッチ処理用の高レベルAPI
     * 
     * @param namesList 複数の名前リスト
     * @param mode 分割モード
     * @param concurrency 並列実行数
     * @return 各リストの分割結果
     */
    suspend fun batchDivide(
        namesList: List<List<String>>, 
        mode: String = "basic",
        concurrency: Int = 3
    ): List<List<DividedName>> = coroutineScope {
        namesList.chunked(concurrency).flatMap { chunk ->
            chunk.map { names ->
                async { divide(names, mode) }
            }.awaitAll()
        }
    }

    /**
     * リソースを解放
     */
    fun close() {
        // HttpClient は自動でリソース管理されるため、特別な処理は不要
    }
}